### ISSUE TITLE
Def Jam: Fight for NY "`PTimer` upsupported" leading to freezing

### ISSUE THREAD

**Issue by Triticum0** [xemu-project/xemu#1124]
### Title

https://xemu.app/titles/45410049/#Def-Jam-Fight-for-NY

### Bug Description

The game never make it past the first loading screen properly and need one of the unimplemented commands to be playable

### Expected Behavior

Should make it to menu

### xemu Version

Version: 0.7.55
Branch: master
Commit: https://github.com/mborgerson/xemu/commit/db389b15086753cb7b72e05b24af205baa742bc3
Date: Sun Jun 26 05:58:57 UTC 2022

### System Information

<html><body>

Field | Value
-- | --
OS | Windows 10
CPU | AMD Ryzen 5 2600 Six-Core Processo
Graphics Device | NVIDIA GeForce RTX 3060 Ti/PCIe/SSE2
Graphics Driver | 4.0.0 NVIDIA 512.95

</body>
</html>

### Additional Context

[xemu.log](https://github.com/mborgerson/xemu/files/8995963/xemu.log)

**Comment by CosmicStardus** [comment]
I'm not sure if this is known, but I made a discovery with this game that I posted in /r/xemu: The demo disc (NTSC-U region) to this game loads just fine in the latest version of xemu (0.7.118). Not only that, it plays great!

Could something simple be stopping the full game from loading?

**Comment by papadkostas** [comment]
@CosmicStardus  could you share more? which exact iso you got?
I'm trying with xemu 0.7.119 usa and eu img and doesn't work, stays for ever in logo.

**Comment by QuesoBlanco12** [comment]
> @CosmicStardus could you share more? which exact iso you got? I'm trying with xemu 0.7.119 usa and eu img and doesn't work, stays for ever in logo.

I can confirm that this issue does not exist in the "Demo" version of this ISO, and the game actually plays just fine.

![image](https://github.com/xemu-project/xemu/assets/76604675/5e3f592f-c4c8-4115-a0f3-c8fb802e0d9e)

If a dev is interested in investigating this, I am willing to provide a working xiso so that this can be investigated further. Just need to know what the best way would be to share it(don't want to break any rules or DMCA's, even if this is a demo disk, just playing it safe unless otherwise told).

**Comment by flex50five** [comment]
Bumping this thread. Anyone been able to get this working with any new updates or bios?

**Comment by MaouYumisu1** [comment]
As of the current build on USA Xemu 0.7.127 result is still the same as above.

**Comment by VamsAndre** [comment]
> > @CosmicStardus could you share more? which exact iso you got? I'm trying with xemu 0.7.119 usa and eu img and doesn't work, stays for ever in logo.
> 
> I can confirm that this issue does not exist in the "Demo" version of this ISO, and the game actually plays just fine.
> 
> ![image](https://private-user-images.githubusercontent.com/76604675/331816036-5e3f592f-c4c8-4115-a0f3-c8fb802e0d9e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwNTYyMDcsIm5iZiI6MTcyMDA1NTkwNywicGF0aCI6Ii83NjYwNDY3NS8zMzE4MTYwMzYtNWUzZjU5MmYtYzRjOC00MTE1LWEwZjMtYzhmYjgwMmUwZDllLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA0VDAxMTgyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI3ZWIzMjQ3MWI2MjBmMzYzMDNlZTlhMDAwMjQyNTAwN2YyZTEzNzdkZDU0NmQ1OGY0ODRmYTI4M2NhMjVkMTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2PEBxx33KlnyU0o6d4edMeIsMZt-iuqFZHtufEz6wcw)
> 
> If a dev is interested in investigating this, I am willing to provide a working xiso so that this can be investigated further. Just need to know what the best way would be to share it(don't want to break any rules or DMCA's, even if this is a demo disk, just playing it safe unless otherwise told).

Maybe it's something like FIFA, in 07 when you package in common ways the iso doesn't start, you have to do it in a specific way or download it from a specific site. It has to do with EA and there must be an iso lock that the demo does not have, it could be the music

**Comment by QuesoBlanco12** [comment]
> > > @CosmicStardus could you share more? which exact iso you got? I'm trying with xemu 0.7.119 usa and eu img and doesn't work, stays for ever in logo.
> > 
> > 
> > I can confirm that this issue does not exist in the "Demo" version of this ISO, and the game actually plays just fine.
> > ![image](https://private-user-images.githubusercontent.com/76604675/331816036-5e3f592f-c4c8-4115-a0f3-c8fb802e0d9e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwNTYyMDcsIm5iZiI6MTcyMDA1NTkwNywicGF0aCI6Ii83NjYwNDY3NS8zMzE4MTYwMzYtNWUzZjU5MmYtYzRjOC00MTE1LWEwZjMtYzhmYjgwMmUwZDllLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA0VDAxMTgyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI3ZWIzMjQ3MWI2MjBmMzYzMDNlZTlhMDAwMjQyNTAwN2YyZTEzNzdkZDU0NmQ1OGY0ODRmYTI4M2NhMjVkMTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2PEBxx33KlnyU0o6d4edMeIsMZt-iuqFZHtufEz6wcw)
> > If a dev is interested in investigating this, I am willing to provide a working xiso so that this can be investigated further. Just need to know what the best way would be to share it(don't want to break any rules or DMCA's, even if this is a demo disk, just playing it safe unless otherwise told).
> 
> Maybe it's something like FIFA, in 07 when you package in common ways the iso doesn't start, you have to do it in a specific way or download it from a specific site. It has to do with EA and there must be an iso lock that the demo does not have, it could be the music

It could very well be.

All I can say is that both versions of the ISOs I have tried are both re-dumps that have been patched the same way to work on xemu.

I tried taking it upon myself to try and seeing what the hiccup could be, but I discovered everything is in Assembly (or at least I think it is, no idea) and unfortunately this makes me as useful as (something very useless here).

I still have both fixed re-dump files available to any developer who would like to investigate this. Thanks!

**Comment by papadkostas** [comment]
Does this even run on a real machine?

**Comment by MaouYumisu1** [comment]
> Does this even run on a real machine?

https://youtu.be/4FEGNnYI8wY?si=PGPhgyePB9MPEDot
If you mean the game, Yes it does. This is played on the original xbox.

**Comment by papadkostas** [comment]
I mean if the actual ISO(not the demo) is even working on the real console? Just to exclude any other problem.

**Comment by PayrollG** [comment]
yeah it never work for me when i play on it i was stuck on the intro how you fix this intro

**Comment by hugobitola** [comment]
Bumping this thread again. I love this game, used to play with my older brother and a cousin for weeks.
We didn't had the original disc, it was an ISO.
Do you need this ISO for investigation?

**Comment by Steff833** [comment]
I Bump too :p. Pisses me off that no one found out since years. Xemu dosnt hang. Maybe it is some sort of copy protection in the games default.xbe. Def Jam behaves nearly similar like burnout3 but with a difference. You can boot burnout 3 via iso file. But you cant load the game via virtual Hdd. For example via a dashboard like xbox media center. Burnout has then also an infinite loading screen. I also tried to load these two games from other dashboards like Avalaunch.

My old real Xbox console dosnt care. It boots both games perfect no matter how. From hdd XBMC or DVD. Btw. I copied and replicated my complete Xbox to Xemu. I Extracted the bios from the Chip, use the same eeprom, the same HDD Key, the same Xbox media Center as an OS. 

Normally (with the same bios) Xemu should now behave exactly like the real hardware. Only def jam and Burnout bitches around.

**Comment by Triticum0** [comment]
Tested on HDD and made no difference.

**Comment by abaire** [comment]
Hey @Triticum0 it looks like the xemu.log you attached expired at some point, could you add a new log? 

I assume from the bug description that there was some info about unimplemented commands, I'm curious what they were. Is it possible that the game is actually running but not updating the display (like Tomb Raider #1120)?

**Comment by Triticum0** [comment]
Here you go  

[output.zip](https://github.com/user-attachments/files/22957879/output.zip)

**Comment by abaire** [comment]
Thanks, I see from that log that it's doing both depth and alpha testing, either of which could result in no fragments being rendered if there was a bug or an incorrect depth comparison. Otherwise nothing jumps out as obvious.

Is anything being displayed while it is in the infinite loading state? If it's not (or if it's just a static image) could you grab a RenderDoc cap when it's in that state so we can see if it's trying to draw something but the updates are being discarded? 

Also, when it's in that state, could you go into the debug monitor and do `info registers` a few times to see if the `EIP` value changes at all? Might need to do it 10+ times if it doesn't seem to be changing (since it might be in a tight loop). If it's not changing, could you paste the output of that command in this thread?

Unfortunately I don't own this one and can't tell if this is just a graphics issue or if it's actually softlocked/deadlocked.

**Comment by Triticum0** [comment]
EIP doesn't  change from 0021f027 and renderdoc only render the logo, don't know cause but also effected official Xenon Fusion also has same behavior where demo works but full game doesn't work so if you got the game + demo disk 36 should be able to compare hardware traces and be able to fix the issue.

Ref: https://consolemods.org/wiki/Xbox_360:Original_Xbox_Games_Compatibility_List
`Stuck on loading but doesn't freeze oddly the demo included with offical xbox magazine game disc 36 works fine with no issue `

Renderdoc

[Def Jam Fight for NY.zip](https://github.com/user-attachments/files/22959960/Def.Jam.Fight.for.NY.zip)

Note: left it running for at least 10 mins.

**Comment by abaire** [comment]
Thanks!

Did you do the renderdoc capture after it had been running for 10 minutes, or right before it froze? If the instruction pointer isn't changing I would've thought that it also would stop populating the pushbuffer, so I'm surprised RenderDoc would capture anything at all.

Could you grab the entire output of `info registers` and paste it here since it seems to be stuck?

**Comment by Joshua-1248** [comment]
[pipeline_state.html](https://github.com/user-attachments/files/23197863/pipeline_state.html)

Hopefully this is everything.

**Comment by abaire** [comment]
Unfortunately a dump of the GL state doesn't bring any new info. 

I'm still looking for the output of `info registers` in the debug monitor to see if there's anything interesting there. If somebody can attach gdb and get a dump of the code around `0x021f027 `that'd also be potentially useful.


This one might require some reverse engineering of the title itself to understand what it's trying to do, so somebody with the disc and the skills to use Ghidra/IDA/whatever will need to be found. I'd be very interested to see if `0x021f027` is inside of `D3D_BlockOnTime` as there's a similar apparent race condition in #1962

**Comment by Triticum0** [comment]
Here you go

CPU#0
EAX=00000008 EBX=00000001 ECX=000023a5 EDX=813d0000
ESI=000023af EDI=0000000a EBP=00234b0c ESP=d0044c74
EIP=0021f027 EFL=00000293 [--S-A-C] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0010 00000000 ffffffff 00cf9300 DPL=0 DS   [-WA]
CS =0008 00000000 80060fff 00c89b00 DPL=0 CS32 [-RA]
SS =0010 00000000 ffffffff 00cf9300 DPL=0 DS   [-WA]
DS =0010 00000000 ffffffff 00cf9300 DPL=0 DS   [-WA]
FS =0020 8004ac8c 00284fff 00c09300 DPL=0 DS   [-WA]
GS =0000 00000000 00000000 00000000
LDT=0000 00000000 0000ffff 00008200 DPL=0 LDT
TR =0018 8005e990 00000068 00008900 DPL=0 TSS32-avl
GDT=     8005eac8 00000038
IDT=     8004af38 00000200
CR0=8001003b CR2=00000000 CR3=0000f000 CR4=00000610
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000 
DR6=ffff0ff0 DR7=00000400
EFER=0000000000000000
FCW=027f FSW=4000 [ST=0] FTW=00 MXCSR=00001fa0
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=0000000000000000 0000000000000000 XMM01=0000000000000000 0000000000000000
XMM02=0000000000000000 0000000000000000 XMM03=0000000000000000 0000000000000000
XMM04=0000000000000000 0000000000000000 XMM05=0000000000000000 0000000000000000
XMM06=0000000000000000 0000000000000000 XMM07=0000000000000000 0000000000000000

**Comment by Triticum0** [comment]
Don't know How long it going to take me to get the debugger working but see if I can get it working today

**Comment by abaire** [comment]
> Don't know How long it going to take me to get the debugger working but see if I can get it working today

It may be faster to use Ghidra with https://github.com/XboxDev/ghidra-xbe if you happen to have it.

**Comment by Triticum0** [comment]
@abaire I check on Ghidra and it is indeed inside D3D_BlockOnTime.

<img width="1640" height="874" alt="Image" src="https://github.com/user-attachments/assets/95236ac5-28bc-4975-bdc1-82e2de2093aa" />

**Comment by abaire** [comment]
Cool, I suspect this is a dup of #1962 then, if it spins in that loop forever. 

It'd be somewhat interesting to see if `SET_FLIP_MODULO` is set to 3 for this game as well (I think this indicates use of triple buffering).

**Comment by abaire** [comment]
@Triticum0 is this reproducible with PGRAPH tracing turned on? If so, I'd be interested in whether it matches the scenario I see w/ Over the Hedge.

If it's like the other bug, turning on the tracing also might change timing sufficiently to avoid the livelock.

**Comment by Triticum0** [comment]
It doesn't =(

**Comment by Triticum0** [comment]
Also check the Ptrace and SET_FLIP_MODULO is set to 3

**Comment by abaire** [comment]
@Triticum0: assuming https://github.com/abaire/xemu/pull/34 works for Over the Hedge, it'd be good to test it on this too to see if it's really the same issue w/ the TV encoder field counter not being populated.

**Comment by Triticum0** [comment]
Tested still stuck on same register address 0021f027

**Comment by abaire** [comment]
Ah, too bad, then this is likely a different issue that may require significant reverse engineering. Unfortunately the current price of the game is far too high.

https://github.com/abaire/xemu/pull/35 has the logging I added to track down #1962
If you could 

1. Run the artifact from that PR and attach the resulting log 
2. Run artifact a second time with PGRAPH tracing enabled (you can hit F9 just before the infinite loading to turn it on) 

I can try to see if there's an obvious error state.

**Comment by ghost** [comment]
I am not en engineer, but could I contribute by buying the game for anyone?

**Comment by Triticum0** [comment]
Here you go

[xemu.zip](https://github.com/user-attachments/files/23528989/xemu.zip)

**Comment by abaire** [comment]
@Triticum0 Thanks! Can you attach your PGRAPH trace as well? Ideally the one from the master build that you captured before or from my `work` branch, I don't need/want the extra info from the special build.

That log is enough to tell me that this is a distinct issue from #1962. This title is still displaying frames and releasing the semaphore whereas that one was stuck waiting for a frame flip. I'm surprised that the EIP would always be in BlockOnTime given those results.

**Comment by Triticum0** [comment]
Here you go

[output.zip](https://github.com/user-attachments/files/23530004/output.zip)

**Comment by abaire** [comment]
Thanks! I can see from this that it's drawing frames with some actual content. Could you get a RenderDoc cap while it's in the infinite loading state?

EDIT: Deleted irrelevant dump. The initial halt was due to unimplemented PTIMER interrupts, see comments below.

**Comment by Triticum0** [comment]
Here you go. Compressed all the captures into 1 file. Also I can send you a copy of the game. as it not that expensive for EU copy of the game. 

[Def Jam Fight for NY Capture 3 Mulltcapture..zip](https://github.com/user-attachments/files/23530685/Def.Jam.Fight.for.NY.Capture.3.Mulltcapture.zip)

**Comment by abaire** [comment]
Thanks! I see that it's just rendering the loading image, so nothing particularly interesting or obviously problematic there.

**Comment by abaire** [comment]
@Triticum0 thanks for the heads up that the PAL game is cheaper, I was able to buy a copy from the UK.

After a lot of reverse engineering, I found that the game is blocked waiting for a kernel event to be set (handle `0x4C`, object address `0xD0008C18`). Testing on HW, this event is set inside of the DPC for the PMC interrupt (the same place that vblank and other interrupts are handled).

The interrupt is looking for the `NV_PMC_INTR_0_PTIMER_PENDING` bit to be set (see https://github.com/XboxDev/nxdk/blob/d0ae96acc926a6c3f6254b289a1a9a3c1db182f1/lib/pbkit/outer.h#L83), but xemu never sets this bit. Further investigation shows that the PTIMER handling was never fully implemented. The writes to `NV_PTIMER_ALARM_0` are received, but nothing ever acts on them.

I guessed at the HW implementation and am able to get further in the loading process before it stalls again on the logo screen. I don't see any further interaction with the PTIMER registers during the new stall; it'll require more debugging/reversing.

**Comment by Triticum0** [comment]
Retested same behaviour as u got. Great job as always.

**Comment by abaire** [comment]
I think it's another race condition; if I test under the debugger with some breakpoints and/or debug print messages I can sometimes get it to pass all of the FMVs and sometimes it won't even start the first FMV. Just to make sure, I patched in the fix for the Conexant field register, no change.

When it stops, I see that all threads are halted at `0x8002189C` which looks like it's called from the usual kernel wait functions (KeWaitForMultipleObjects/KeWaitForSingleObject amongst others). I confirmed that there is no pending alarm, so this is presumably due to some new wait condition. On the occasions where I can get execution to proceed, I see that it does continue to set alarms, so it's probably not an issue with rollover behavior in the PR.

Unfortunately `0x8002189C` is in a non user dispatch level so XBDM can't break there and give me the thread context; but I can break at the end of the method in gdb and for at least one of the threads it was called from `KeDelayExecutionThread`.

Edit: I also see at least one thread in `KeWaitForMultipleObjects` and another in `KeWaitForSingleObject`
Edit: `0x8002189C` looks like it's an artifact of how XBDM stops threads and is a red herring.

**Comment by abaire** [comment]
I wrote a qemu plugin to assist in debugging this: https://github.com/abaire/xemu-plugin-wait-trace
It requires some extensions to xemu to work (it allows arbitrary "plugin" commands to be sent via the debug monitor): https://github.com/abaire/xemu/tree/debug/adds_plugin_debug_monitor_interaction

It looks like there are at least a few threads that get stuck in `KeWaitForSingleObject`. This may or may not be an actual problem.

```
[WaitTrace] Wait table:
	KeWaitForSingleObject
		371927 obj: d0055fd8 $esp: d00abc40
		372024 obj: d005c5d8 $esp: d0122c58
		374049 obj: d0008318 $esp: d00249f8
```

Reinforcing the theory that this is a race condition, with this plugin running it is noticeably harder to reproduce the issue. With the plugin I can get into the game more than half the time, with a clean build of #2568 and no plugins it freezes on one of the FMVs (very often the Def Jam trademark/licensing message just after the initial load screen) much more often than it works.


EDIT: Some more investigation:
```
Thread 56	#9352	{d0063938}
Thread 160	#9454	{d0061358}
```

It looks like threads 56 and 160 lock up first, after which other threads start to lock up. Looking at historical calls to KeSetEvent, it seems that thread 12 (looks like the render thread, many D3D interactions) is typically responsible for releasing the locks on these threads. Thread 12 is still active and is actually rendering the stuck screen.

It may also be Thread 8 that is responsible for the locks; it is also not deadlocked. Thread 8 is in a loop that continually waits on a handle until the value at `0x003c9624` is set to 0, it looks like this would exit the thread and may never happen in normal operation either. The loop increments two counters: `0x003c9684` and `0x003c9688`. When in the frozen state for a little while, both were set to 0x1FF0 (and incremented indefinitely). These counters are read from many places.

Possible backtrace:
```
> /gbt 12
EIP: 0x1ece25
# 0      0x11d47 in Game!.text [ 0x1ece20 ]
# 0      0x1150a in Game!.text [ 0x11cd0 ]
# 0      0x1f6344 in Game!.text [ 0x11500 ]
# 0      0x1f68ac in Game!.text [calc]
#  1      0xb001a2c6 in xbdm.dll!.text ? 0xb00239fa ?


> /gbt 56
EIP: 0x8002189c
#  1      0x80025f07 in xboxkrnl.exe!.text ? 0x80022f5a ?
#   2      0x800218ce in xboxkrnl.exe!.text ? 0x80025939 ?
#    3      0x8002dc2a in xboxkrnl.exe!.text ? 0x8001c05c ?
#     4      0x8002d45a in xboxkrnl.exe!.text ? 0x80022f5a ?
#      5      0x8001ad54 in xboxkrnl.exe!.text ? 0x8002c7bc ?
# 0      0x1f5bd7 in Game!.text [calc]
#       6      0x1f6066 in Game!.text ? 0x1f5dda ?
#       6      0x1edd7f in Game!.text ? 0x1f6057 ?
#       6      0x1e4beb in Game!.text ? 0x1edd70 ?
# 0      0x1ed0b4 in Game!.text [calc]
# 0      0x1f68ac in Game!.text [calc]
#        7      0xb001a2c6 in xbdm.dll!.text ? 0xb00239fa ?


> /gbt 160
EIP: 0x8002189c
#  1      0x80025f07 in xboxkrnl.exe!.text ? 0x80022f5a ?
#   2      0x800218ce in xboxkrnl.exe!.text ? 0x80025939 ?
#    3      0x8002dc2a in xboxkrnl.exe!.text ? 0x8001c05c ?
#     4      0x8002d45a in xboxkrnl.exe!.text ? 0x80022f5a ?
#      5      0x1f6066 in Game!.text ? 0x1f5dda ?
#      5      0x1edd7f in Game!.text ? 0x1f6057 ?
#      5      0x77215 in Game!.text ? 0x1edd70 ?
# 0      0x1ed0b4 in Game!.text [calc]
# 0      0x1f68ac in Game!.text [calc]
#       6      0xb001a2c6 in xbdm.dll!.text ? 0xb00239fa ?


> /gbt 8
EIP: 0x1f6066
# 0      0x1edd7f in Game!.text [ 0x1f6057 ]
# 0      0x1ed5fd in Game!.text [ 0x1edd70 ]
# 0      0x1ed0a5 in Game!.text [calc]
# 0      0x1f68ac in Game!.text [calc]

```

**Comment by abaire** [comment]
After looking at the backtraces I found them to be suspiciously similar to the scenario that led me to find the missing PTIMER alarm interrupts in the first place. I went ahead and wrote a simple test case for the alarm behavior and found two things:
1) There is no magic value to disable the alarm.
2) The alarms fire every time the low 32 bits of the NV2A clock wrap around, as long as the alarm interrupt is enabled.

My initial implementation got both of these wrong; since I saw this game setting the alarm constantly I assumed it was a one-shot timer. I guess the kernel sets the alarm to `0xFFFFFFFF` just to have a known state rather than to disable things.

Both of these are fixed now in PR #2568 and I've been unable to reproduce the freeze after a couple dozen restarts. It'd be great if somebody could verify that the artifact attached to the PR fixes the freeze for them as well.

**Comment by Triticum0** [comment]
Well done abaire. Shame that it took while to get primer work correctly. I Test any game still freezes + ptrace incase use it In logs. Hope you take a break and enjoy Christmas as only week half away. Seen all work in background, you deserve it. 😀☃️

**Comment by PayrollG** [comment]
Yeah it’s good when it fix every time I get on main menu it froze I reset few times it keep freezing more time all I gotta y’all gotta fix the freeze and stuff and y’all should be good

**Comment by retchless** [comment]
Tested it again with the latest code changes in the PR, and sadly, it still freezes at random moments. Glad to see some action though!

**Comment by abaire** [comment]
There are no new code changes; I don't plan to do anything else with this particular PR unless it becomes clear that the freezes are due to the PTIMER code specifically. 

Next time it freezes, please get a dump of the registers (see comment above) and add it here; hopefully that will provide a hint as to whether it's waiting forever on an alarm or is some other problem.

**Comment by retchless** [comment]
Does this help @abaire 

```
# info registers

CPU#0
EAX=fffffffb EBX=00236398 ECX=00236398 EDX=0023641c
ESI=fd000000 EDI=00236398 EBP=00000000 ESP=800395b4
EIP=00223772 EFL=00000206 [-----P-] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0010 00000000 ffffffff 00cf9300 DPL=0 DS   [-WA]
CS =0008 00000000 8003cfff 00c89b00 DPL=0 CS32 [-RA]
SS =0010 00000000 ffffffff 00cf9300 DPL=0 DS   [-WA]
DS =0010 00000000 ffffffff 00cf9300 DPL=0 DS   [-WA]
FS =0020 80035bdc 0027cfff 00c09300 DPL=0 DS   [-WA]
GS =0000 00000000 00000000 00000000
LDT=0000 00000000 0000ffff 00008200 DPL=0 LDT
TR =0018 8003b160 00000068 00008900 DPL=0 TSS32-avl
GDT=     8003b298 00000038
IDT=     80035e60 00000200
CR0=8001003b CR2=00000000 CR3=0000f000 CR4=00000610
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000 
DR6=ffff0ff0 DR7=00000400
EFER=0000000000000000
FCW=027f FSW=4020 [ST=0] FTW=3f MXCSR=00001fa0
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=fa00000000000000 4008 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=c000000000000000 4016
XMM00=3ec971933ec97193 3ec971933ec97193 XMM01=c36bc6b2c36303ac c3728624c387bf47
XMM02=427c102841a7d750 c28f96fcc33706a8 XMM03=43aab9624398ab2b 43750a584318066c
XMM04=43e1712a43ae12c2 43a5f10d43a843fa XMM05=3e8000003f000000 3f40000000000000
XMM06=3e8000003f000000 3f40000000000000 XMM07=3000000030000000 3000000030000000

```

**Comment by abaire** [comment]
Unfortunately it doesn't tell me too much. I believe thats inside of an ISR related to graphics processing.

The register I'm mostly interested in is EIP and ESP. If you can get a bunch of logs while it's frozen and paste just the EIP & ESP value pairs we can hopefully get an idea of what it's doing. It will be tricky as there are multiple threads and most of them are probably stuck in wait loops. Sorry for not being more exact in my previous message!

If you're technically inclined, you can use the plugin I wrote to do the initial diagnosis in https://github.com/xemu-project/xemu/issues/1124#issuecomment-3619622368 - this will allow you to get some stack trace guesses for the broken frames but it's a relatively complicated process.

If there's any way to reliably trigger the freeze, please add it here; eventually I'll look into it again if nobody figures it out in the meantime.

**Comment by retchless** [comment]
@abaire I am technically inclined, so I will attempt to get the plugin going. 

That said, it's reproducible super easily. Open Def Jam Fight for NY. If it doesn't freeze during the initial cut scene, go into battle, choose a couple of random characters to fight against each other, start the battle. It almost always freezes during the map selection or during the cut scene loading the fight.

FYI, before the pull request (#2568), the game wouldn't even load.  Now it's just freezing during cut scenes.

**Comment by abaire** [comment]
I was able to reproduce a freeze w/ some instrumentation around the `PTIMER` code.

From what I can see, the game disabled the PTIMER interrupt just before the freeze, it then never reenabled it.
I see that the PFIFO is still running and PGRAPH is still running through frames.

My initial guess was that I have a mistake in how I deal with the interrupt being disabled (currently the timer used to fire the alarm is deleted as soon as the game disables the interrupt). A brute force attempt to work around this by simply not deleting the timer and allowing the alarm to always fire did not resolve the problem.

It is particularly odd that the game never seems to disable the PTIMER interrupt except just before freezing. The freeze is not in a consistent place, so there is some race condition/bug/etc... that must trigger the interrupt disable. Presumably either this is not supposed to happen at all or there is some recovery that isn't triggered (perhaps similar/related to the missing timer irq that the PR attempts to fix).

More investigation will be needed to trace this down.


Edit:

It does seem likely to be related to PTIMER. The interrupt is only disabled in a few specific conditions, one of which is when the game logic fails to find any registered timers. I suspect that there's a race condition where a timer can be registered slightly in the past. The current code handles this by bumping the high bits of the alarm up by 1, which theoretically means the timer is set about four seconds into the future. The high bits are inaccessible to the guest software, so this in itself shouldn't be a problem.

Immediately firing the IRQ when a timer is set into the past seems to prevent the freeze but leads to regular hitching of ~4 seconds (likely a wait in the guest code for the high time to be incremented).

I'll need to update the tests to see if I can verify how HW handles an alarm set in the past. Since there's no way to set the high bits of the alarm it seems odd that it'd immediately fire (if the low bits are close to the max any alarm set meaningfully in the future would have to wrap around to a lower value)


Looking at the actual values used for the alarms, they're always around 0xFE5413 ticks apart, which is very suspiciously close to 16.67 ms (1/60 of a second). Mapped along with the emulated PTIMER time value, the values drift apart until the registered alarm rolls over many read/update cycles before the emulated GPU clock value catches up. This doesn't seem right to me; I'd expect the title to adjust the alarm time based on the actual time it reads back.

Edit 2:
I extended the HW tests and verified that, in a simple case at least, xemu behavior with this PR mirrors HW behavior. I do not see the divergence observed in the game.

Further debug logging has shown that there is a latency spike just before the game softlocks, here's a partial log:

```
PTIMER 0x420 (ALARM_0) = 0x70D3E026
Schedule ptimer alarm 53432541216 - now 53416005504 = 16536523 ns
READ PTIMER TIME_1: 0x410 = 0xC
READ PTIMER TIME_0: 0x400 = 0x6FD800C0
READ PTIMER TIME_1: 0x410 = 0xC
PTIMER alarm fired after 16665184 ns. Was scheduled for 16536523 ns
Schedule ptimer alarm 57727508512 - now 53432556672 = 4295163124 ns

PTIMER 0x100 (INTR_0) = 0x1
READ PTIMER TIME_1: 0x410 = 0xC
READ PTIMER TIME_0: 0x400 = 0x70D5D5E0
READ PTIMER TIME_1: 0x410 = 0xC

PTIMER 0x420 (ALARM_0) = 0x71D22549
Schedule ptimer alarm 53449205056 - now 53432678656 = 16527210 ns
READ PTIMER TIME_1: 0x410 = 0xC
READ PTIMER TIME_0: 0x400 = 0x70D65EA0
READ PTIMER TIME_1: 0x410 = 0xC
PTIMER alarm fired after 51742464 ns. Was scheduled for 16527210 ns  // this looks bad, timer is way late
Schedule ptimer alarm 57744172352 - now 53484301120 = 4260080793 ns

PTIMER 0x100 (INTR_0) = 0x1
PTIMER alarm fired after 4259893440 ns. Was scheduled for 4260080793 ns
Schedule ptimer alarm 62039139648 - now 57744190560 = 4295160373 ns

PTIMER 0x100 (INTR_0) = 0x1

PTIMER 0x140 (INTR_EN_0) = 0x0
Debugger disconnected
```

I'm under the impression that the `QEMU_CLOCK_VIRTUAL` should be unaffected by the speed of execution (i.e., slowdowns due to inefficient emulation should be invisible to the guest), so I'm not sure what would explain a huge delay in the qemu timer going off.

It seems that `QEMU_CLOCK_VIRTUAL` is tied to the host clock unless qemu is run with `-icount` (which precludes use of KVM)

Edit 2026-02-05:

I think the clock is a red herring. Here's another fail case:

```
PTIMER 0x420 (ALARM_0) = 0x8AD044BF
Schedule ptimer alarm 2328904864 - now 2312357280 = 16548394 ns
READ PTIMER TIME_1: 0x410 = 0x0
READ PTIMER TIME_0: 0x400 = 0x89D417A0
READ PTIMER TIME_1: 0x410 = 0x0
PTIMER alarm fired after 16669184 ns. Was scheduled for 16548394 ns
  Late by 120790 ns
Schedule ptimer alarm 6623872160 - now 2328924448 = 4295158997 ns

PTIMER 0x100 (INTR_0) = 0x1
READ PTIMER TIME_1: 0x410 = 0x0
READ PTIMER TIME_0: 0x400 = 0x8AD5D8A0
READ PTIMER TIME_1: 0x410 = 0x0

PTIMER 0x420 (ALARM_0) = 0x8BCF2A0A
Schedule ptimer alarm 2345609728 - now 2329289440 = 16321089 ns
READ PTIMER TIME_1: 0x410 = 0x0
READ PTIMER TIME_0: 0x400 = 0x8AD71120
READ PTIMER TIME_1: 0x410 = 0x0
PTIMER alarm fired after 16705184 ns. Was scheduled for 16321089 ns
  Late by 384095 ns
Schedule ptimer alarm 6640577024 - now 2345630624 = 4295157686 ns

PTIMER 0x100 (INTR_0) = 0x1
READ PTIMER TIME_1: 0x410 = 0x0
READ PTIMER TIME_0: 0x400 = 0x8BD16BC0
READ PTIMER TIME_1: 0x410 = 0x0

PTIMER 0x420 (ALARM_0) = 0x8CCCE2A4
Schedule ptimer alarm 2362237600 - now 2345765632 = 16472777 ns
READ PTIMER TIME_1: 0x410 = 0x0
READ PTIMER TIME_0: 0x400 = 0x8BD1F080
READ PTIMER TIME_1: 0x410 = 0x0
PTIMER alarm fired after 16628160 ns. Was scheduled for 16472777 ns
  Late by 155383 ns
Schedule ptimer alarm 6657204896 - now 2589819616 = 4067585369 ns

PTIMER 0x100 (INTR_0) = 0x1

// Why is this not checking TIME_1/TIME_0/TIME_1 and setting a new alarm?


PTIMER alarm fired after 4294963712 ns. Was scheduled for 4067585369 ns
  Late by 227378343 ns
Schedule ptimer alarm 10952172192 - now 6657220512 = 4295162966 ns

PTIMER 0x100 (INTR_0) = 0x1

PTIMER 0x140 (INTR_EN_0) = 0x0
Debugger disconnected

Process finished with exit code 0
GDB Server stopped, exit code 137
```

Here we can see that the final alarm (0x8CCCE2A4) fired on time but for some reason the title did not schedule a new alarm. Reversing the interrupt handler implies that this would happen if the 2-alarm queue is empty and the callback associated with the timer does not schedule a new alarm. I am not sure yet which of these two conditions (or both) is true and where the alarms are generally scheduled. If this can fail even with an on-time alarm, it seems that the slow fires could be a symptom of the problem that leads things down the softlock path.

Debugging further, it seems that the early exit path at 0x00216456 ends up being taken just before the softlock, preventing a new alarm from being registered. Just before this early exit, there's a `rdtsc` and a check to verify that the target time for the callback being scheduled is in the future, bailing early (without calling the callback) if the alarm is scheduled in the past.

I think this title might be particularly sensitive to latency that would never exist on the HW. I suspect that our qemu timer fires slightly too slow (for whatever contention/etc... reason) and the `rdtsc` reported time has been allowed to advance such that our interrupt comes too late. Then the title fails to register a new alarm and when the stale alarm re-fires due to wraparound it enters an irrecoverable state where alarm interrupts are entirely disabled.

**Comment by abaire** [comment]
More evidence that the issue here is timing related.

I patched the method at `0x00216410` to print out the target time passed to the function and the current `rdtsc` result. Just before the softlock you can see that these cross over, the tsc time is less than the target time:

```
src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 12 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x737463B6 0x4  RDTSC 0x72C7B4AC 0x4  PTIMER 0x53AB9B20 0x5

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 12 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x742EE2C4 0x4  RDTSC 0x73C30CF1 0x4  PTIMER 0x55025500 0x5

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 12 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x74E961D2 0x4  RDTSC 0x7438E3DF 0x4  PTIMER 0x55A34620 0x5

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 12 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x75A3E0E0 0x4  RDTSC 0x7F4C41C7 0x4  PTIMER 0x64BD3FA0 0x5
```
Before the failure, timers are scheduled at significantly variable points in the future:
`0xACAF0A`
`0x6BD5D3`
`0xB07DF3`
and the trigger in the past
`-0x9A860E7`

If I run with `icount` this never happens (as far as I can tell, I let the game run for ~30 minutes. Generally it softlocks within 4 in the non `icount` case); furthermore the values are always much further apart:

```
src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 96 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0xA6021E38 0x5  RDTSC 0xB51AE0C9 0x2  PTIMER 0xFFEA08A0 0x2

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 96 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0xA6BC9D46 0x5  RDTSC 0xB7174565 0x2  PTIMER 0x29F5140 0x3

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 96 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0xA7771C54 0x5  RDTSC 0xB7D0A22E 0x2  PTIMER 0x39C1260 0x3

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 96 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0xA8319B62 0x5  RDTSC 0xB88B4801 0x2  PTIMER 0x49A9440 0x3
```

It is possible that they are further apart due to the time it takes me to install the hooks, but that time is similar across the `icount` and non-instruction counting case.

In the icount case, the delta between the target and the RDTSC stay very similar, as do the deltas between the components across calls.

Nest step is to see why the target varies so wildly between icount and non-icount. It does not seem to be pinned to the PTIMER clock.





----

For the sake of completeness, here's the output from HW

```
src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 28 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x273BA705 0x123  RDTSC 0x26815815 0x123  PTIMER 0xC3B504E0 0x1

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 28 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x27F62613 0x123  RDTSC 0x273BD722 0x123  PTIMER 0xC4B35220 0x1

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 28 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x28B0A521 0x123  RDTSC 0x27F655E2 0x123  PTIMER 0xC5B19E60 0x1

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 28 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x296B242F 0x123  RDTSC 0x28B0D408 0x123  PTIMER 0xC6AFEA20 0x1

src/xbox/debugger/xbdm_debugger.cpp:517    (01) info    > [DEBUGGER] 
DebugStr: thread_id: 28 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x2A25A33D 0x123  RDTSC 0x296B550F 0x123  PTIMER 0xC7AE3A20 0x1
```

Timers are scheduled consistently around `0xBA4Exx` (about 12ms) into the future

**Comment by Triticum0** [comment]
What is icount is that enabled in code or a build compiler flag. I probably test a custom build with the feature implemented,  If performance is acceptable with no debug flags, I leave it in a new issue, once ptimer merged as you got it pretty well implemented at this point. As don't how useful it is doing anymore debugging other issue with softlocks might have similar timing issues. There are only 5 games that freeze consistently 2 them being Regression not including games which freeze depending on hardware(ninja gaiden). So those game might help fix some timing issues. I updated my timing issue https://github.com/xemu-project/xemu/issues/1099 and add game that freeze and softlock now its becoming an issue.

**Comment by abaire** [comment]
> What is icount is that enabled in code or a build compiler flag. 

https://www.qemu.org/docs/master/system/qemu-manpage.html it's a command-line flag

I am dubious that unpredictability in the clock would explain things in #1099, those are likely unrelated. I would not add softlocks to that bug, I'd keep it about things running faster/slower than hardware and leave lockups in per-title issues as they likely need case-by-case debugging.

**Comment by abaire** [comment]
I added alarm tracking to my [clock test](https://github.com/abaire/nxdk_low_level_nv2a_tests/blob/main/src/clock_main.cpp)

If I run without `icount` and put my machine under significant load (rebuilding xemu from scratch while watching videos while running xemu with a bunch of extra stderr output streaming to my gdb and then dragging the xemu window around) I can get alarms to fire up to 34 milliseconds late. This increases my suspicion that wall-clock delays are bleeding over into the virtual clock and triggering the softlock.

Edit: Even with my machine under standard load, if I double click to make xemu fullscreen I'll get 16ms+ lag spikes.

**Comment by abaire** [comment]
mborgerson pointed out that launching xemu with the `-d nochain` option might improve the consistency of interrupt handling. It wasn't enough to prevent the freeze in my instrumented debug build but might be worth checking on a release build of PR #2568 

I suspect it will still freeze for some combination of machine, and it will almost certainly freeze if you interact with the xemu UI (dragging windows, toggling fullscreen) but it might be enough to be playable in the short term.

---

It seems like the interrupt just fires too late sometimes, from the guest's perspective. Anecdotally, the delay before a failure seems to be about double the intended delay; seemingly one frame's worth of interrupt is missed.

To capture some state from the disassembly:

`0x000d7de0: void __thiscall FUN_000d7de0__ScheduleAlarm?(Thread80Class *this)`  - this function uses the current QPC time to decide how far into the future to schedule the alarm. It checks the current time - the last alarm time against the last requested interval and uses that as a modifier against the precalculated ticks_per_period (0x00ba7f0e = 12,222,222, or about 16.67 ms @ 733 MHz).

At the end of the function it calls `FUN_00216410_CheckRDTSCAndSetPTIMERAlarm`

`0x00216410: HRESULT? FUN_00216410_CheckRDTSCAndSetPTIMERAlarm(uint TargetTimeLow,uint TargetTimeHigh,void *Callback,void *UserData)` - this function invokes `rdtsc` to check the current time against the target time. If the target is in the past, it bails out immediately with an error code. The alarm scheduler assumes this failure can never happen (in HW it should be impossible) and this allows execution to proceed witihout rescheduling the PTIMER alarm. The old alarm fires again after wraparound (~4 seconds later) and the ISR code sees that it received an unscheduled alarm and turns off alarm interrupts, leaving the title in an irrecoverable state where it's waiting for an alarm that was never scheduled and could never be received even if it was.

---

The numbers on the xemu side all look reasonable. Here is another fully instrumented event series after which the guest fails to register a new timer:

```
PTIMER 0x420 (ALARM_0) = 0xDC2FE559
	Schedule ptimer alarm now virt 6508430000 qpc 4772848664 ptimer: 3694126400 - now 3678590816 = 15536344 ns
	PTIMER alarm fired at 6523970000 virt 4784244664 qpc after 16664192 ns ptimer. Was scheduled for 15536344 ns ptimer
		Late by 1127848 ns (ptimer) 1 ms
	PTIMER ALARM FLAGGED at 6523983000 virt 4784254197 qpc
	Schedule ptimer alarm now virt 6523986000 qpc 4784256397 ptimer: 7989093696 - now 3694146048 = 4295158933 ns
PCI INT FIRED at 6523996000 virt 4784263731 qpc: delta since alarm raised virt 13000 ns 0 ms

PTIMER 0x100 (INTR_0) = 0x1 time 6524081000 virt 4784326064 qpc since alert fire (virt) 98000 ns 0 ms
```


---

Edit 2026-02-10

I instrumented things a bit more on the xemu side and it looks like (in the error case) there's a massive gap between the time the guest clears the interrupt and the time it reads the timer (about a frame's worth of delay).

E.g.,:

```
PTIMER 0x100 (INTR_0) = 0x1 time 0x6 0xF543966F qpc since alert fire (virt) 118000 ns 0 ms
READ PTIMER TIME_0: 0x400 = 0xE0D83FA0 at 0x6 0xF663E74A qpc
```

While there is a case in the guest disassembly that appears to wait on a vblank, a breakpoint set there is never hit.

**Comment by abaire** [comment]
I did some more debugging and confirmed that the failure case is always after a ~1 frame delay. There is a place in the timer handler that would await a vblank, but it is never taken.

This made me suspicious that the DPC itself might be getting interrupted during handling of the alarm event (after the interrupt flag is cleared but before it invokes the callback to register a new timer). That reminded me of #1962, there is code in the vblank handler that will wait an extra frame if the TV encoder chip's even/odd frame tracking does not match the CPU's expectation. I have a pending fix for this in PR #2537 

My initial test of merging PR #2537 with the PTIMER alarm fix PR #2568 seems positive. I was able to consistently get in game and let my character get beaten without a freeze (before I could pretty reliably freeze within the first 5 minutes from title boot).

This will need more testing to verify. I've created a branch merging the two in my fork, https://github.com/abaire/xemu/actions/runs/21878955286 will have the artifacts.

**IMPORTANT** this will still fail if you do certain things w/ the xemu UI (e.g., toggle fullscreen, maybe resize the window or drag it around). I believe this is a separate issue w/ the way xemu allows the virtual clock to run while it's holding the BQL; it will need an additional fix.

@retchless @Triticum0 (and anybody following this thread) would you be able to check to see if this merged build performs any better for you?

**Comment by Triticum0** [comment]
I Tested it briefly and im still getting freezing issue with both of them being merged.  I think also freeze if sharers haven't been cached.

**Comment by abaire** [comment]
> I Tested it briefly and im still getting freezing issue with both of them being merged. I think also freeze if sharers haven't been cached.

That's an interesting observation; does it still freeze if you disable shader caching? I've only tested on Linux and macOS, neither of which caches shaders.

I think ideally any time spent compiling/saving shaders would be done while the virtual timer is paused, same as interacting with the xemu window.

**Comment by Triticum0** [comment]
> That's an interesting observation; does it still freeze if you disable shader caching? I've only tested on Linux and macOS, neither of which caches shaders.

 disabling shader cache still exhibited frequent freezing behaviour on bazzite 43

**Comment by abaire** [comment]
> disabling shader cache still exhibited frequent freezing behaviour on bazzite 43

Thanks for checking! 

@Triticum0 Out of curiosity, do you have vsync enabled? 

I did when performing tests on macOS, but I wonder if a host-side delay for refresh could end up pausing long enough to trigger the problem. It'd be interesting to test with vsync on and off to see if there's any difference in behavior.

I'll eventually instrument the frame counter handling in the game to see if it's triggering a delay despite the odd/even fix. It's also possible that there's a check elsewhere that would account for this, but I'm increasingly suspicious that the delay leading to the softlock always seems to be ~16ms longer than prior working delays.

**Comment by Triticum0** [comment]
yes I was using vsync but with it off still froze =(

**Comment by abaire** [comment]
I noticed that I forgot to rebase these fixes on top of the change that moves the UI vblank wait to a separate thread.

@Triticum0 could you test the artifact from https://github.com/abaire/xemu/actions/runs/22050428959 when it finishes?

My suspicion is still that something is delaying 1 frame before fully processing the alarm. I've instrumented (what I believe is) the vblank handler and the alarm handler in the game code and so far I haven't found evidence that it's waiting which makes me continue to worry that the guest clock is left running while the host is doing something. I want to rule out the vblank wait in xemu.c; moving it to a separate thread seems like a good thing to double check. If that doesn't fix anything, I might see if using `cpu_disable_ticks` around likely host-side stalls has any effect.

**Comment by Triticum0** [comment]
I tested it with v-sync and cache shaders turned off and still froze on that  build.

**Comment by abaire** [comment]
> I tested it with v-sync and cache shaders turned off and still froze on that build.

Thanks for checking.

I added some more instrumentation and caught a failure that inexplicably took more than 1 frame, causing the handler of the alarm scheduled in this block to attempt to schedule a new alarm in the past due to the lag here:

```
DebugStr: thread_id: 136 text: DPC at 0x9 0x6E29EB9A

DebugStr: thread_id: 136 text: Service timer interrupt at 0x9 0x6E3241B8

DebugStr: thread_id: 136 text: MyCheckRDTSCAndSetPTIMERAlarm Target 0x9 0x6ED18087 RDTSC 0x9 0x6E600022 PTIMER 0x0 0x576505A0
DebugStr: thread_id: 136 text: > CTX: TicsPerP: 0x00000000 0x00BA7F0E  QPCStart: 0x00000009 0x6E5F6572  DT: 0x00000000 0x00721B15  QPCFreq: 0x00000000 0x2BB5C755

DebugStr: thread_id: 136 text: Calling SetNVPTIMERAlarm from CheckRDTSCAndSetPTIMERAlarm: 0x9 0x6E6C54E8 rdtsc

// Huge delta since the interrupt started at 0x6E3241B8 : 83,222,333 ns

DebugStr: thread_id: 136 text: Service timer interrupt exit at 0x9 0x732820F5
```

I print a message anytime the DPC for the interrupt handler is entered. This DPC handles vblank, timers, pfifo, etc... As expected, the DPC is called exactly once during the period that the timer interrupt is handled.

I print a message at the entry to the timer specific handling code called by the DPC. As we can see, it happens promptly after the DPC is entered (the DPC loops over various interrupts, so it could do other work before handling the timer).

I print a message when the timer interrupt checks the current TSC clock and decides to schedule an alarm in the future and when the alarm is actually scheduled, verifying that there is no wait logic being triggered once the TSC check passes.

Another case with a significant, inexplicable jump in guest clock:

```
[@0x002239a0] DPC at 0x4 0xC97FA22E
[@0x002232e0] Service timer interrupt at 0x4 0xC982FD8E

[@0x000d7e8a] MyCheckRDTSCAndSetPTIMERAlarm 
      Target 0x4 0xCA36113D 
      RDTSC 0x4 0xC98F8306 
      PTIMER 0x0 0xE82CEB20
> CTX: TicsPerP: 0x00000000 0x00BA7F0E  QPCStart: 0x00000004 0xC98F5253  DT: 0x00000000 0x00A6BEEA  QPCFreq: 0x00000000 0x2BB5C755

[@0x00216473] Calling SetNVPTIMERAlarm from CheckRDTSCAndSetPTIMERAlarm: 0x4 0xC992D013 rdtsc

[@0x002233d8] Calling GetFullPTIMERTimeNS from ServiceTimerInterrupt: 0x4 0xC9961A43 rdtsc

[@0x002233fb] Service timer interrupt exit at 0x4 0xC998DAF3

DPC at 0x4 0xC9B3FAB3




DPC at 0x4 0xCA3A5B7B

Service timer interrupt at 0x4 0xCA3E5469

// Significant stall: 0xCB0273F6 - 0xCA3E5469 = 12,853,133
// Note that there is no call to `GetFullPTIMERTimeNS` reported here, which means that the loop in the service
// func has not skipped the callback that calls CheckRDTSCAndSetPTIMERAlarm. The clock has just jumped ~1 frame
// ahead in time.

MyCheckRDTSCAndSetPTIMERAlarm 
      Target 0x4 0xCAF0904B 
      RDTSC 0x4 0xCB0273F6 
      PTIMER 0x0 0xEA26AD40
> CTX: TicsPerP: 0x00000000 0x00BA7F0E  QPCStart: 0x00000004 0xCB024621  DT: 0xFFFFFFFF 0xFFEE4A2A  QPCFreq: 0x00000000 0x2BB5C755
```

**Comment by PayrollG** [comment]
Sheesh

**Comment by abaire** [comment]
> Sheesh

Please feel free to mute/unsubscribe from this thread, this is a tricky problem that is taking a lot of work to debug. Besides testing, I don't think any other user input is needed at this point so there's no reason for non-contributors to be inundated with notifications.

@Triticum0 can you test this build when you have some time? https://github.com/abaire/xemu/actions/runs/22087775169 

I attempted to disable the guest clock progression while the main loop is locked for xemu UI work.

**Comment by Triticum0** [comment]
@abaire  Retested crashed at same point of last build which tend to be when just making in game before able to control fighter.

Also update to lasted build there audio issues in that build which have been fixed by matt in masters as it effected all ea games. https://github.com/xemu-project/xemu/pull/2741

**Comment by abaire** [comment]
Thanks for trying. I'll eventually look to see if there are other laggy operations that could benefit from having the clock paused but at this point I don't have any good ideas left. This game just seems to be unusually sensitive to timing. On my M3 MacBook it regularly gets into the game and will run for at least 10-15 minutes, so it could be something like shader compilation or just CPU speed that accounts for the difference.


> Also update to lasted build there audio issues in that build which have been fixed by matt in masters as it effected all ea games. [#2741](https://github.com/xemu-project/xemu/pull/2741)

That build was already rebased on top of `master` so it should have the latest code as of yesterday.

**Comment by Triticum0** [comment]
@abaire Should be fixed with #2945 haven't tested yet but was most likely what was causing the issue

**Comment by abaire** [comment]
> [@abaire](https://github.com/abaire) Should be fixed with [#2945](https://github.com/xemu-project/xemu/pull/2945) haven't tested yet but was most likely what was causing the issue

I'd be surprised if an APU fix resolves this problem; I did quite a bit of reverse engineering of the game and am pretty confident it requires PTIMER alarm support. 

That said, please let me know if it really does resolve it without my PTIMER alarm PR so I can drop that change. Since the freeze is heavily timing dependent it'll need quite a bit of testing to verify that it doesn't just happen to work sometimes.

**Comment by Triticum0** [comment]
I did some test and your right, it didn't change alarm desync behavior with host and guest timings and does still freezes with 2 fix merged together.
